okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java

```diff
@@ -600,6 +600,32 @@ public void run() {
         // ClientFrameHandler need to be started after connectionPreface / settings, otherwise it
         // may send goAway immediately.
         executor.execute(clientFrameHandler);
+
+        executor.execute(new Runnable() {
+          @Override
+          public void run() {
+            synchronized (lock) {
+              lock.notify();
+            }
+          }
+        });
+
+        long waitStartTime = System.currentTimeMillis();
+        synchronized (lock) {
+          try {
+            lock.wait(1000);
+          } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+          }
+        }
+
+        long waitEndTime = System.currentTimeMillis();
+        if (waitEndTime - waitStartTime >= 1000) { // never got notified
+          startGoAway(0, ErrorCode.INTERNAL_ERROR, Status.UNAVAILABLE
+              .withDescription("Handshake timed out due to insufficient threads"));
```

Comment:
Let's have the error message say what happened instead of assuming what happened. It could be that the machine is overloaded and just really slow. "Timed out waiting for second handshake thread. The transport executor pool may have run out of threads."